source/slang/glsl.meta.slang

```diff
@@ -2825,21 +2825,1724 @@ public uint rayQueryGetIntersectionTypeEXT(rayQueryEXT q, bool committed)
     return 0;
 }
 
+// TODO: implementation of built-in variables; proper tests; these are stubs
+// likley related to the following issue since GLSL adds new 
+// 'system' variables: https://github.com/shader-slang/slang/issues/411
+
+__generic<T : __BuiltinType>
+[require(glsl)]
+void typeRequireChecks_shader_subgroup_GLSL() {
+    // the following is a seperate function call, since else the `__requireGLSLExtension` and associated __intrinsic_asm is ignored if the calling function also calls an __intrinsic_asm
+    __target_switch
+    {
+    case glsl:
+        if (__type_equals<T, half>()
+            || __type_equals<T, float16_t>()
+            ) __requireGLSLExtension("GL_EXT_shader_subgroup_extended_types_float16");
+        else if (__type_equals<T, uint8_t>()
+            || __type_equals<T, int8_t>()
+            ) __requireGLSLExtension("GL_EXT_shader_subgroup_extended_types_int8");
+        else if (__type_equals<T, uint16_t>()
+            || __type_equals<T, int16_t>()
+            ) __requireGLSLExtension("GL_EXT_shader_subgroup_extended_types_int16");
+        else if (__type_equals<T, uint64_t>()
+            || __type_equals<T, int64_t>()
+            ) __requireGLSLExtension("GL_EXT_shader_subgroup_extended_types_int64");
+
+        __intrinsic_asm "";
+    }
+}
+
+__generic<T : __BuiltinType>
+[require(spirv)] [require(glsl)]
+void shader_subgroup_preamble() {
+    // checks needed for shader_subgroup functions; __requireGLSLExtension does not work 
+    // (does not add the ext specified correctly to the compile output; using extended type
+    // will result in error for using the type)
+    __target_switch
+    {
+    case glsl:
+        typeRequireChecks_shader_subgroup_GLSL<T>();
+    case spirv:
+        spirv_asm {};
+    }
+
+} 
+
+// GL_KHR_shader_subgroup_basic Built-in Variables 
+
+__glsl_extension(GL_KHR_shader_subgroup_basic) [require(glsl)] 
+public uint gl_NumSubgroups;
```

Comment:
We need to turn these into properties and define the SPIRV intrinsics for them.

```
[require(glsl)] [require(spirv)]
public property uint gl_NumSubgroups
{
     get
     {
           __target_switch
          {
          case glsl:
                  __requireGLSLExtension("GL_KHR_shader_subgroup_basic");
                  __intrinsic_asm "gl_NumSubgroups";
           case spirv:
                  return spirv_asm {
                       result:$$uint = OpLoad builtin(NumSubgroups:uint);
                    };
          }
     }
}
```

Reply:
has the following feature been tested outside of defining in a struct/related-type?

I've been finding that `public property uint gl_NumSubgroups` as a standalone does not really work?

Reply:
You probably need to modify `isDeclAllowed` to allow a PropertyDecl to be nested under ModuleDecl at slang-parser.cpp:4215.

Reply:
propertiesDecl does have permission; 

the error which was provided seems to be an internal compiler related subject(?): `error 99999: Slang compilation aborted due to an exception of class Slang::InternalError: unimplemented: addInArg case`

Reply:
Let me know if you need help on debugging the error. You can search the souce files with the given error messages and set a break point there. And when you reproduce the issue the program will stop at the breakpoints, and you can trace backward to figure out what condition made the program to branch into the error message printing code. That condition will be the real reason why you were getting the error message.

Don't be scared by words like "internal error" or "unimplemented". Those words actually mean "developers need to debug this with a debugger attached".